src/ACL.cpp

```diff
@@ -141,36 +141,69 @@ QFlags< ChanACL::Perm > ChanACL::effectivePermissions(ServerUser *p, Channel *ch
 	bool write    = false;
 	ChanACL *acl;
 
+	// Iterate over all parent channels from root to the channel the user is in (inclusive)
 	while (!chanstack.isEmpty()) {
 		ch = chanstack.pop();
-		if (!ch->bInheritACL)
+		if (!ch->bInheritACL) {
 			granted = def;
+		}
 
 		foreach (acl, ch->qlACL) {
 			bool matchUser  = (acl->iUserId != -1) && (acl->iUserId == p->iId);
 			bool matchGroup = Group::appliesToUser(*chan, *ch, acl->qsGroup, *p);
+
+			bool applyFromSelf   = (ch == chan && acl->bApplyHere);
+			bool applyFromParent = (ch != chan && acl->bApplySubs);
+
+			// Not using applyFromAny will ignore the context flags for the current ACL
+			bool applyFromAny = applyFromSelf || applyFromParent;
+
+			// Denying traverse is special, because denying it for any parent channel
+			// implicitly means all sub channels are entirely inaccessible as well.
+			// However, we still want to make sure that unchecking "applies to
+			// this channel" for the parent channel is honored.
+			bool applyDenyTraverse = applyFromParent || acl->bApplyHere;
+
 			if (matchUser || matchGroup) {
-				if (acl->pAllow & Traverse)
+				// These do not grant or deny anything. We merely
+				// check, if we are missing traverse AND write in this
+				// channel and therefore abort without any permissions later on.
+				if (applyFromAny && acl->pAllow & Traverse) {
 					traverse = true;
-				if (acl->pDeny & Traverse)
+				}
+				if (applyDenyTraverse && acl->pDeny & Traverse) {
 					traverse = false;
-				if (acl->pAllow & Write)
+				}
+				if (applyFromAny && acl->pAllow & Write) {
 					write = true;
-				if (acl->pDeny & Write)
+				}
+				if (applyFromAny && acl->pDeny & Write) {
 					write = false;
-				if (ch->iId == 0 && chan == ch && acl->bApplyHere) {
-					if (acl->pAllow & Kick)
+				}
+
+				// These permissions are only grantable from the root channel
+				// as they affect the users globally. For example: You can not
+				// kick a client from a channel without kicking them from the server.
+				if (ch->iId == 0 && applyFromSelf) {
+					if (acl->pAllow & Kick) {
 						granted |= Kick;
-					if (acl->pAllow & Ban)
+					}
```

Comment:
```suggestion
					granted |= (acl->pAllow & Kick);	
```
This would be more compact but I have the feeling the readability would suffer :thinking:

Reply:
I agree that readability would suffer and rather leave it as is. Going through this function and trying to understand it was hard enough already :smiling_face_with_tear: